### ISSUE TITLE
Loading items for Folders with '#' or '%' by CamlQuery doesn't work

### ISSUE THREAD

**Issue by bajce** [pnp/pnpcore#839]
Im loading the Items by CamlQuery. Query looks like this

`<View>
    <ViewFields>
        <FieldRef Name='FileRef'/>
        <FieldRef Name='FileDirRef'/>
    </ViewFields>
    <Query>
        <Where>
            <Eq> <FieldRef Name='FileDirRef'/> <Value Type='Text'>{fileDirRef}</Value></Eq>
        </Where>
    </Query>
</View>`

and im using following code to load the Items:

`  list.LoadItemsByCamlQuery(new CamlQueryOptions()
                        {
                            ViewXml = queryXml,
                            DatesInUtc = true,
                            PagingInfo = nextPage,
                            FolderServerRelativeUrl = folderServerRelativeUrl 
                        },
                        item => item.FieldValuesAsText);`

Unfortunately no items are returned if Folder has '#' or '%' in it's name.

I tried different approaches to escape `fideDirRef` value in query as well as FolderServerRelativUrl value in CamlQueryOptions but non of them helped.

**Comment by jansenbe** [comment]
Thanks for reporting this @bajce , I'll have a look and try to reproduce and fix

**Comment by jansenbe** [comment]
@bajce : did you try to escape the folder using XML escape techniques. Any textual content inserted into a CAML query should be properly escaped like shown here: 

```xml
<Eq><FieldRef Name="Title" /><Value Type="Text"><![CDATA[SharePoint&InfoPath]]></Value></Eq>
```

Depending on the use case the following approach can also work, just escape the special chars for XML: so for example replace the & with &amp;

**Comment by bajce** [comment]
@jansenbe Items containing XML special characters can be escaped using `<![CDATA[ ]]` or `SecurityElement.Escape()` and loading items of folders with XML special character using both of approaches works as expected, but # and % are however a valid XML characters (e.g `SecurityElement.Escape("SharePoint#InfoPath")` won't change the string at all). But I tried both approaches and none of them was working for '#' and '%'. 

Maybe this has something to do with the URL special characters. Characters like # and % can often be misunderstood in URL ( A hash sing for page fragment, and percent is used for escaping special characters).

I also tested to build navigate to specific folder in browser like this `.../site/X/Y/Folder #` and doesn't work as well. By escaping the URL special characters (%23 instead of #) I was able to open the folder path in browser. 
I also tried this approach for CAML Query as well es for `CamlQueryOptions#FolderServerRelativeUrl` but that didn't help.

**Comment by jansenbe** [comment]
@bajce : can you try replacing `#` with `&#35;` and `%` with `&#37;` and see if that works for you?

**Comment by bajce** [comment]
@jansenbe Thanks for the code snippet. I tested that and unfortunately im still not getting the items. I tried couple different approaches: using method from above, replacing # and % manually and I tried that by only escaping it in CAML query as well as CAML query and `FolderServerRelativeUrl `.

**Comment by jansenbe** [comment]
@bajce , did test this myself and I'm getting an approach to work by adapting the CAML query (see code below) to query on `FileDirRef`. Using the FolderServerRelativeUrl property is not (yet) working for me, feels like this is a bug in SharePoint which you might want to raise via an official support channel.

```csharp
var myList = await context.Web.Lists.GetByTitleAsync("CAML");

await myList.LoadItemsByCamlQueryAsync(new CamlQueryOptions()
{
    ViewXml = "<View Scope='Recursive'><ViewFields><FieldRef Name='Title' /></ViewFields><Query><Where><Eq><FieldRef Name='FileDirRef'/><Value Type='text'>/sites/prov-2/Lists/CAML/special&#37;folder&#35;here</Value></Eq></Where></Query></View>",
    DatesInUtc = true,
    //FolderServerRelativeUrl = "/sites/prov-2/Lists/CAML/special%folder#here"
});
```

The list setup is this one:
![image](https://user-images.githubusercontent.com/7451219/167404244-5302b8f0-ba1c-4b39-9ae5-a8051c19806c.png)

**Comment by bajce** [comment]
@jansenbe thank you for providing working code sample. I was always testing with `FolderServerRelativeUrl` . 
Using your approach im getting the items but not all, (Folder were missing), but by changing the Scope from `Recursive` to `RecursiveAll` I was able to get all items. Thank you on that.

Also I did some further tests and its seams like only problem is when `FolderServerRelativeUrl` is set and folder name contains `#` or `%`, Even without replacing the characters I was able to get items.

This solution works for me:
When I build Caml query I escape the folder name by using `SecurityElement.Escape(fileDirRef)`. It escapes all XML Special characters (Not `#` and `%`, but its not necessary as I mentioned  above).

In order to keep query as performant as possible I set scope to `RecursiveAll` only if `fileDirRef` contains `#` or `%` and in that case im not setting `FolderServerRelativeUrl`, otherwise `FolderServerRelativeUrl` is set to `fileDirRef` and no `Scope` is set in Caml query.

@jansenbe Thank you for your help!
